Code/GraphMol/FindRings.cpp

```diff
@@ -1105,105 +1099,84 @@ int symmetrizeSSSR(ROMol &mol, VECT_INT_VECT &res) {
     nsssr = findSSSR(mol, sssrs);
   } else {
     sssrs = mol.getRingInfo()->atomRings();
     nsssr = rdcast<unsigned int>(sssrs.size());
   }
 
-  VECT_INT_VECT_CI srci;
-  INT_VECT copr;
-  for (srci = sssrs.begin(); srci != sssrs.end(); srci++) {
-    copr = (*srci);
-    res.push_back(copr);
+  for (const auto& r: sssrs) {
+    res.emplace_back(r);
   }
 
   // now check if there are any extra rings on the molecule
   if (!mol.hasProp(common_properties::extraRings)) {
-    // no extra rings nothign to be done
+    // no extra rings nothing to be done
     return rdcast<int>(res.size());
   }
   const VECT_INT_VECT &extras =
       mol.getProp<VECT_INT_VECT>(common_properties::extraRings);
 
-  // std::cerr<<" extras "<<extras.size()<<std::endl;
   // convert the rings to bond ids
-  VECT_INT_VECT bsrs, bextra;
-  RingUtils::convertToBonds(sssrs, bsrs, mol);
-  RingUtils::convertToBonds(extras, bextra, mol);
-  INT_VECT munion, nunion, symids;
-  Union(bsrs, munion);
-  INT_VECT sr, exr;
-  INT_VECT_CI eri;
-  unsigned int eid, srid, ssiz;
-  unsigned int next = rdcast<unsigned int>(bextra.size());
-  // now the trick is the following
-  // we will replace each ring of size ssiz from the SSSR with
-  // one of the same size rings in the extras. Compute the union of of the new
-  // set
-  // if all the union elements of the new set if same as munion we found a
-  // symmetric ring
-  for (srid = 0; srid < nsssr; srid++) {
-    sr = bsrs[srid];
-#if 0
-        std::cerr<<"  consider: "<<srid<<std::endl;
-        std::copy(sssrs[srid].begin(),sssrs[srid].end(),std::ostream_iterator<int>(std::cerr," "));
-        std::cerr<<"  | ";
-        std::copy(sr.begin(),sr.end(),std::ostream_iterator<int>(std::cerr," "));
-        std::cerr<<std::endl;
-        std::cerr<<"------"<<std::endl;
-#endif
-    ssiz = rdcast<unsigned int>(sr.size());
-    INT_VECT exrid;
-    exrid.push_back(srid);
-    Union(bsrs, nunion, &exrid);
-    for (eid = 0; eid < next; eid++) {
-      // if we already added this ring continue
-      // FIX: if the ring has already been added,it probably shouldn't be
-      // in the list at all?  Is this perhaps the most efficient way?
-      if (std::find(symids.begin(), symids.end(), static_cast<int>(eid)) !=
-          symids.end()) {
+  VECT_INT_VECT bondsssrs;
+  RingUtils::convertToBonds(sssrs, bondsssrs, mol);
+
+  //
+  // For each "extra" ring, figure out if it could replace a single
+  // ring in the SSSR. A ring could be swapped out if:
+  //
+  // * They are the same size
+  // * The replacement doesn't remove any bonds from the union of the bonds
+  //   in the SSSR.
+  //
+  // The latter can be checked by determining if the SSSR ring is the unique
+  // provider of any ring bond. If it is, the replacement ring must also
+  // provide that bond.
+  //
+  // May miss extra rings that would need to swap two (or three...) rings
+  // to be included.
+
+  // counts of each bond
+  std::unordered_map<int, int> bondCounts;
```

Comment:
wouldn't this be faster if we used an `std::vector<int>` that's initialized to 0?

Reply:
I was thinking about systems like proteins, for which there are few rings but many atoms.

But I think you're right. Always good to use vector.

Reply:
btw, this sped it up by about 50%.  Now this diff takes symmetrizeSSSR from 5s to about 0.001s for the nanotube.